torch/csrc/jit/passes/graph_fuser.cpp

```diff
@@ -494,19 +493,24 @@ struct GraphFuser {
     bool any_changed = true;
     while (any_changed) {
       any_changed = false;
-      for (auto it = nodes.rbegin(); it != nodes.rend();) {
+      for (auto it = block->nodes().rbegin(); it != block->nodes().rend();) {
         bool changed;
         std::tie(it, changed) = scanNode(*it);
         any_changed |= changed;
       }
     }
+    for (Node * node : block->nodes()) {
+      for (Block * sub_block : node->blocks()) {
+        GraphFuser(sub_block).run();
```

Comment:
What stops nodes from fusing across blocks? Fusing across blocks is not always safe since it may try to move a producer into a consumer that is only executed conditionally (or executed multiple times, which is a perf bug). 
I believe we need to add a check into all `isFusable` functions (and other places where we move instructions around) that the two operations are in the same blocks.

Reply:
Hmmm yeah, I think I checked that it was safe in BatchMM, and then concluded that I'm happy with "every block-level function has independent state, and works on nodes from a single block, so it should be ok", but yeah, it's obviously false when we're trying to traverse the graph via inputs.